site/app/controllers/forum/ForumController.php

```diff
@@ -705,183 +708,191 @@ public function markPostUnread(): JsonResponse {
+        $post = $thread->getPosts()->filter(function ($x) use ($post_id) {
+            return $x->getId() === $post_id;
+        })->first();
 
-        if (!$this->core->getAccess()->canI("forum.modify_post", ['post_author' => $post['author_user_id']])) {
+        if ($post === false) {
+            return $this->core->getOutput()->renderJsonFail("Unable to find post : {$post_id}");
+        }
+        if (!$this->core->getAccess()->canI("forum.modify_post", ['post_author' => $post->getAuthor()->getId()])) {
                 return $this->core->getOutput()->renderJsonFail('You do not have permissions to do that.');
         }
-        if (isset($_POST['thread_id']) && $post['thread_id'] !== intval($_POST['thread_id'])) {
-            return $this->core->getOutput()->renderJsonFail("You do not have permission to do that.");
+        if ($post->getThread()->isLocked() && !$this->core->getUser()->accessAdmin()) {
+            return $this->core->getOutput()->renderJsonFail('Thread is locked');
         }
-        if (isset($_POST['edit_thread_id']) && $post['thread_id'] !== intval($_POST['edit_thread_id'])) {
-            return $this->core->getOutput()->renderJsonFail("You do not have permission to do that.");
+
+        $status_edit_thread = true;
+        if ($thread->getFirstPost()->getId() === $post->getId()) {
+            $status_edit_thread = $this->editThread($thread);
         }
-        if (!empty($_POST['edit_thread_id']) && $this->core->getQueries()->isThreadLocked($_POST['edit_thread_id']) && !$this->core->getUser()->accessAdmin()) {
-            return $this->core->getOutput()->renderJsonFail('Thread is locked');
+        $status_edit_post = $this->editPost($post);
+
+        $message = [];
+        if (!$status_edit_thread) {
+            $message[] = "Thread update failed.";
+            $this->core->getCourseEntityManager()->refresh($thread);
         }
-        elseif (!empty($_POST['thread_id']) && $this->core->getQueries()->isThreadLocked($_POST['thread_id']) && !$this->core->getUser()->accessAdmin()) {
-            return $this->core->getOutput()->renderJsonFail('Thread is locked');
+        if (!$status_edit_post) {
+            $this->core->getCourseEntityManager()->refresh($post);
+            $message[] = "Post update failed.";
         }
-        elseif ($modify_type == 0) { //delete post or thread
-            $thread_id = $_POST["thread_id"];
-            $thread_title = $this->core->getQueries()->getThread($thread_id)['title'];
-            if ($this->core->getQueries()->setDeletePostStatus($post_id, $thread_id, 1)) {
-                $type = "thread";
-            }
-            else {
-                $type = "post";
+        if (count($message) > 0) {
+            return $this->core->getOutput()->renderJsonFail(join(" ", $message));
+        }
+        if (!$thread->isChanged() && !$post->isChanged()) {
+            return $this->core->getOutput()->renderJsonFail("No data submitted. Please try again.");
+        }
```

Comment:
What is the reasoning to have this? You have to add quite a bit of logic to each entity to track this for potentially little payoff.

Reply:
This functionality exists already on main, so I'm preserving it. The tracking is easy to hide behind setters, and they show up again later when preparing a success response for the user.

Reply:
Where is this on main?

Reply:
Line 794-795 in the red diff

Reply:
Thanks for pointing it out. As far as I can tell, the original functionality does not check that there was a change to the data. It only checks if there was a successful update to the database regardless of state change. Having the `is_changed` member variable in adding more complexity than needed in the entities and can easily be missed in future changes.

Reply:
The other use case for `is_changed` is right below this, when building the response to the user. I think these several checks justify the tracking and doing so through setters is straightforward enough it should be easy to maintain. Plus this check will force anyone adding new edit functionality to track those changes lest their testing fail.

If we get rid of change tracking, we'll have to change what notifications get sent and what the user sees. Do you have any thoughts on what to do here?

Reply:
> we'll have to change what notifications get sent and what the user sees

Which ones are you referring to?

Reply:
The remainder of this function (lines 771-810) deals with formatting a notification, sending a websocket message, sending the formatted notification, and then returning to the user. The different cases (thread and post both changed, only thread changed, only post changed) would have to be joined into one if we remove `is_changed`.

Notably we would not be able to choose whether to send 'edit_post' or 'edit_thread' socket messages. We could decide to refresh the whole thread each time a post is edited, but getting that to behave correctly with freshly edited posts would require a significant js refactor as well.

Reply:
So from a UI perspective, it looks like the functionality is as follows:

* Edit the top post (the original post in the thread) in any way (content, title, etc) => an email is sent saying the thread was updated and lists the title and body (even if just the title was updated)
* Edit any other post => an email is sent saying the post was updated

So in the first case, the code always chooses to notify that both the thread and post were updated since the database query succeeds. So to keep current functionality (not author's intended functionality), you would want to match those cases which do not require tracking if the entity has changed.